factcast-snapshotcache-mongodb/src/main/java/org/factcast/core/snap/mongo/MongoDbSnapshotCache.java

```diff
@@ -0,0 +1,225 @@
+
+  // Recommended by the docs to use Majority for gridfs operations
+  public static final TransactionOptions txnOptions =
+      TransactionOptions.builder()
+          .readPreference(ReadPreference.primary())
+          .readConcern(ReadConcern.MAJORITY)
+          .writeConcern(WriteConcern.MAJORITY)
+          .build();
+
+  private final MongoDbSnapshotProperties properties;
+  private final MongoCollection<Document> collection;
+  private final MongoClient mongoClient;
+
+  @VisibleForTesting @Setter private GridFSBucket gridFSBucket;
+
+  public MongoDbSnapshotCache(
+      @NonNull MongoClient mongoClient,
+      @NotNull String databaseName,
+      @NonNull MongoDbSnapshotProperties properties) {
+    this.mongoClient = mongoClient;
+    this.properties = properties;
+    MongoDatabase database = mongoClient.getDatabase(databaseName);
+    this.collection = database.getCollection("factus_snapshot");
+
+    // Grouped by projection, but only contains documents that have an aggregateId set
+    String compoundIndexResult =
+        collection.createIndex(
+            Indexes.ascending(PROJECTION_CLASS_FIELD, AGGREGATE_ID_FIELD),
+            new IndexOptions().partialFilterExpression(Filters.exists(AGGREGATE_ID_FIELD)));
+    log.debug(
+        "Create compound index on factus_snapshot collection returned: {}", compoundIndexResult);
+
+    // Second index is for the projection class only, which is used for snapshots that are not
+    // related to an aggregate
+    String aggregateIdIndexResult =
+        collection.createIndex(Indexes.ascending(PROJECTION_CLASS_FIELD));
+    log.debug(
+        "Create aggregate ID index on factus_snapshot collection returned: {}",
+        aggregateIdIndexResult);
+
+    // Third index for TTL management of documents, expires the document after 0 seconds of the
+    // `expireAt` field. So, immediately.
+    collection.createIndex(
+        Indexes.ascending(EXPIRE_AT_FIELD), new IndexOptions().expireAfter(0L, TimeUnit.SECONDS));
+
+    // Create the GridFS bucket with helper class to store binaries
+    gridFSBucket = GridFSBuckets.create(database);
+    // Warmup the gridFS bucket by doing a dummy upload and delete (Its forbidden later in txns)
+    byte[] tiny = new byte[] {0};
+    ObjectId tmp = gridFSBucket.uploadFromStream("_warmup_", new ByteArrayInputStream(tiny));
+    gridFSBucket.delete(tmp); // cleanup
+  }
+
+  @Override
+  public @NonNull Optional<SnapshotData> find(@NonNull SnapshotIdentifier id) {
+    Document query = createQueryById(id);
+    Document result = collection.find(query).first();
+
+    if (result == null) {
+      return Optional.empty();
+    }
+
+    // Field ID of the binary in GridFS
+    ObjectId fileId = result.getObjectId(FILE_ID_FIELD);
+
+    byte[] bytes;
+    // Stream the GridFS file to a destination (here, a file on disk)
+    try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+      gridFSBucket.downloadToStream(fileId, out);
```

Comment:
out of curiosity, is there a default timeout for these gridFS CRUD operations or is there none by default? Not saying we should change anything here, maybe you already considered/discussed this